src/service.rs

```diff
@@ -704,24 +709,19 @@ where
 
     /// Open the handle corresponding to the protocol
     #[inline]
-    fn protocol_open(
-        &mut self,
-        id: SessionId,
-        proto_id: ProtocolId,
-        address: SocketAddr,
-        ty: SessionType,
-        remote_pubkey: &Option<PublicKey>,
-        version: &str,
-    ) {
+    fn protocol_open(&mut self, id: SessionId, proto_id: ProtocolId, version: &str) {
         debug!("service session [{}] proto [{}] open", id, proto_id);
-
-        let session_context = SessionContext {
-            id,
-            address,
-            ty,
-            remote_pubkey: remote_pubkey.clone(),
-            version: version.to_owned(),
-        };
+        {
+            let session_context = self
+                .session_contexts
+                .get_mut(&id)
+                .expect("Protocol open without session open");
```

Comment:
This situation is theoretically impossible, and the entire event flow guarantees this.